docs/proposals/local-cluster-ux.md

```diff
@@ -0,0 +1,184 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubernetes Local Cluster Experience
+
+This proposal attempts to improve the existing local cluster experience for kubernetes.
+The current local cluster experience is sub-par and often not functional.
+This proposal is a RFC!
+
+## Primary Goals
+
+From a high level the goal is to make it easy for a new user to run a Kubernetes cluster and play with curated examples that require least amount of knowledge about Kubernetes.
+These examples will only use kubectl and only a subset of Kubernetes features that are available will be exposed.
+
+- Works across multiple OSes - OS X, Linux and Windows primarily.
+- Single command setup and teardown UX.
+- Unified UX across OSes
+- Minimal dependencies on third party software.
+- Minimal resource overhead.
+- Eliminate any other alternatives to local cluster deployment.
+
+## Secondary Goals
+
+- Enable developers to use the local cluster for kubernetes development.
+
+## Non Goals
+
+- Simplifying kubernetes production deployment experience
+- Supporting all possible deployment configurations of Kubernetes like various types of storage, networking, etc.
+
+
+## Local cluster requirements
+
+- Includes all the master components & DNS (Apiserver, scheduler, controller manager, etcd and kube dns)
+- Basic auth
+- Service accounts should be setup
+- Kubectl should be auto-configured to use the local cluster
+- Tested & maintained as part of Kubernetes core
+
+## Existing solutions
+
+Following are some of the existing solutions that attempt to simplify local cluster deployments.
+
+### [Spread](https://github.com/redspread/spread)
+
+Spread's UX is great!
+It is adapted from monokube and includes DNS as well.
+It satisfies almost all the requirements, excepting that of requiring docker to be pre-installed.
+It has a loose dependency on docker.
+New releases of docker might break this setup.
+
+### [Kmachine](https://github.com/skippbox/kmachine)
+
+Kmachine is adapted from docker-machine.
+It exposes the entire docker-machine CLI.
+It is possible to repurpose Kmachine to meet all our requirements.
+
```

Comment:
On Sat, Apr 16, 2016 at 12:57:19AM -0700, runseb wrote:
> @rata just saw your comments, sorry for the delay.
> 
> Even though it is a toy model, I very often spawn docker-machines on cloud
> providers. For the simple fact that it is faster (to start and download
> images).

Well, it might be faster or not. But there are very simple ways to run on a
cloud provider. I don't see why running on a cloud provider should be a goal of
"minikube".

> 
> As for validity of things locally. I was not referring to the yaml specs and
> such, but to networking environment, security settings etc.

I don't see what in a one node cluster can break ("Also when doing something
purely local you tend to make assumptions that are not valid remotely, and when
you move to a cloud your stuff breaks.") when moving to the cloud.

Containers inside a pod communicate via localhost just fine (locally and in the
cloud), service name will work locally just fine (so will in the cloud),
networking when using a cloud provider "just works", etc.